### DISCUSSION TITLE
how to run code after form with file upload is submitted?

### DISCUSSION THREAD

**Discussion by scottw-finao** [filamentphp/filament#14039] [Help]
### Package

Form builder

### Package Version

v3

### How can we help you?

I have a form that is uploading a font file for a tool I'm building. I need to take whatever file is uploaded and run a couple of format conversions on it. (I want a woff version for displaying it on forms and need an svg version for the tool to use)

I was able to figure out how to use the state change to extract the correct font name for filling some of the form fields, but now I'm stuck trying to figure out how to execute code after I click the 'create' button.  I assumed it was afterFormValidated() but I can't seem to find somewhere to put that which is a) supported, b) actually runs when I click create.

I generated my Resource using artisan make:filament-resource --generate FontResource and it created number of pages. With some poking through source code I was able to get a break point to stop by creating a override for afterValidate() in the CreateRecord page, but I'm not sure this is the right place to put the code and can't seem to find any examples of the right/recommended way to access data from the forms.  And since there is apparently a hook somewhere for afterFormValidated, I'd rather learn where that is supposed to be run instead.

The docs on the website seem to give some good examples, but repeatedly I find they don't give enough information on precisely where or in what context this code is supposed to run.  Before I give up on filament entirely for poorly documented tools, can someone point me in the right direction?

**Comment by scottw-finao** [comment]
I also seem to see a 'mutateFormDataBeforeCreate()' but I'm unsure how to get the  TemporaryUploadedFile data or something similar in this context so I can know the absolute path of the uploaded file.

**Comment by e2d2-dev** [comment]
Hi,

on the FileUpload::make('yourFile') add

`
->afterStateUpdated(function (FileUpload $component, Component $livewire, Forms\Set $set, ?TemporaryUploadedFile $state)
`

then You should have all Objects You need. To understand further how You can use this just have a look on the integrated ImportAction of Filament or to be more specific the CanImportRecords trait how it can be done.


ciao ed

**Comment by scottw-finao** [comment]
Going through the source code, I found a couple of functions that seem to do the trick. One of them didn't even appear to be documented anywhere. I wanted these things to happen primarily on a 'create' and not necessarily on an 'edit' so I lookedat the CreateRecord object.  There was a 'mutateFormDataBeforeCreate()' that allowed me to change some of the data coming back from the form itself, then I found an 'afterCreate()' that would allow me to manipulate the uploaded file only after the new record was created.

These both seem to do what I need, but I'm still unclear how the file upload interacts with these as I couldn't seem to find any populated objects in the 'afterCreate()' that would give me details about the upload.  I was able to get a string from the 'uploadedfile' field so I implemented a 'saveUploadedFileUsing()' method on the form's upload field so I could make sure the file was saved somewhere that would correlate with the value for uploaded file:

```
FormComponents\FileUpload::make('uploadedfile')
                    ->label('Font File')
                    ->live(true)
                    ->acceptedFileTypes($types)
                    ->preserveFilenames()
                    ->disk('svgfontdir')
                    ->directory('uploaded')
                    ->afterStateUpdated(function (Get $get, Set $set, ?string $old, TemporaryUploadedFile $state) {
                        // get the original uploaded file's name, pull the fontid out of it for the local stored name
                        // set that fontid in a hidden field to be stored in the database.
                        $pi = pathinfo($state->getClientOriginalName());  // get the basename of the original filename
                        $sanitized = sanitize_string($pi['filename']) . '.' . $pi['extension']; // clean basename per filesystem requirements
                        // #01a - store the sanitized original filename in a placeholder hidden field
                        $set('filename', $sanitized);
                        // I give a friendly name suggestion field 'fontname' that uses same method as fontid
                        // #02a - this may be redundant here as I might be able to set fontId and fontname at this point
                        if(empty($get('fontname'))) // get the actual id for the font out of the uploaded file for local naming
                            $set('fontname', get_fontid_from_fontfile($state->getRealPath()));
                    }),
                    ->saveUploadedFileUsing(function (TemporaryUploadedFile $file,): void {
                        $filename = $file->getClientOriginalName();
                        $extension = $file->getClientOriginalExtension();
                        $fullpath = $file->getRealPath(); // included solely for debugging/inspection
                        // clean original filename to save the uploaded file in an 'uploaded' directory to store original font long term
                        $filename = sanitize_string(str_replace('.' . $extension, '', $filename)) . '.' . $extension;
                        $fsm = new FilesystemManager(app());
                        $disk = $fsm->disk('svgfontdir');
                        $disk->putFileAs('uploaded', $file, $filename);
                    })
```

Then my overrides in CreateFont.php:

```
<?php

namespace Treii28\SvgtextFilament\Resources\FontResource\Pages;

use Filament\Resources\Pages\CreateRecord;
use Illuminate\Filesystem\FilesystemManager;
use Livewire\Features\SupportFileUploads\TemporaryUploadedFile;
use Treii28\SvgtextFilament\Resources\FontResource;

class CreateFont extends CreateRecord
{
    protected static string $resource = FontResource::class;
    protected function afterCreate(): void
    {
        // NOTE: I couldn't seem to find a property or object with the upload details at this point
        // formats to convert the file too
        $save_formats = ['ttf', 'otf', 'woff', 'woff2', 'svg'];
        $font = $this->record;
        $ext = pathinfo($font->getAttribute('uploadedfile'), PATHINFO_EXTENSION);
        unset($save_formats[array_search($ext, $save_formats)]); // remove uploaded version from create list
        // copy original format to appropriate directory with font-id as filename
        $fsm = new FilesystemManager(app());
        $disk = $fsm->disk('svgfontdir');
        $copyto = sprintf('%s/%s.%s', $ext, $font->getAttribute('fontid'), $ext);
        $disk->copy('uploaded/'.$font->getAttribute('uploadedfile'), $copyto);

        // convert remaining formats using converTo()
        //   function uses original uploaded version as input and saves result into appropriate subdirectories
        foreach($save_formats as $format) {
            $font->convertTo($format);
        }
    }
    protected function mutateFormDataBeforeCreate(array $data): array
    {
        $form = $this->form;
        // #01b - move temporary 'filename' value from placeholder in form back into the 'uploadedfile' Model property
        if(array_key_exists('filename', $data) && !empty($data['filename'])) {
            $data['uploadedfile'] = $data['filename'];
            unset($data['filename']);
        }
        // #02b - possibly redundant here
        $data['fontid'] = get_fontid_from_fontfile(svgtext_fonts('uploaded/'.$data['uploadedfile']));
        return $data;
    }
}

```

Some of that is a little redundant as I have yet to refactor it. I was still figuring out the best places to do each step so I'm deriving the sanitized fontid in more than one place at present but will eventually pull out the redundancies.

I would be curious to know, however, if overriding 'afterCreate' is a good idea or not or if there is a better way to do it, perhaps in the same form scheme in the Resource file as a sub-Method.

Also, I am still trying to figure out how to manipulate the form on edits.  This method seems to blank out the uploadedfile value if I edit other fields in the form without uploading a new file. And I'm still unsure what happens when I do upload a different file.  My preference would be to have 'edits' give a text field or even a hidden field for the 'uploadedfile' value after the initial upload with perhaps an option to change the field to an upload-field if desired.  But I can probably figure some of this out later.

At this point I'm trying to look into including either a modal or a folded portion of the form for a relationship on the font files.  Converting fonts to the svg form I need for my tool using font-forge doesn't always produce consistent results so I include a 'translation' record on every font that includes values for scaling the size and moving the default position of the font vertically and horizontally so it better aligns in the center of the SVG output.  I'll probably create an HTML5 tool to allow visual moving and scaling of a sample for setting these translations, but first need to figure out how I can handle it in the form itself.
